docs/Memory Model.md

```diff
@@ -148,7 +148,76 @@ follows:
 
 ### Function Pointers
 
-This section is TBD, and will be filled in as part of the work on function pointers.
+Function pointers are in common usage in LLVM, making it crucial that Hieratika is able to support
+their usage. To that end, the following section outlines their integration with the memory model.
+
+The precondition for this design is to first understand that we are not considering function
+pointers to be _true pointers_ at this stage. While LLVM insists that they _are_—and offsets from
+them can be used to access adjacent data such as
+[prefix](https://llvm.org/docs/LangRef.html#prefix-data) or
+[prologue](https://llvm.org/docs/LangRef.html#prologue-data) data—these features do not seem to be
+used by Rust and hence can be safely ignored for now.
+
+As a result, Hieratika treats function pointers _specially_, as follows:
+
+- In LLVM IR, function pointer values are _always_ derived from function pointer constants,
+  referring directly to the encoded name of some function.
+- The Hieratika compiler, at the point of generating function stubs, will generate global constants
+  for these function pointer constants, derived from the correct `blockaddress` expression,
+  containing a _relocatable_ reference to a block (`block_ptr`), and information about which module
+  the block was defined in (`module_id`).
+- The compiler then generates _dispatch functions_ (named
+  `format!("__hieratika_dispatch_local_{function_type}")`). These dispatch functions take the
```

Comment:
The only thing I think this really needs is for us to formalize the type mangling scheme -- specifically, we can say "it's the same as used in polyfills" or we can standardize on one of the mangling schemes used by things like C++ or a representation scheme like python's `struct`.

While most mangling schemes are super gross, the use of them is necessary if we want to work with some standard tooling at any point -- various evolved-from-c-tools limit the practical length of a symbol string to pretty short (31, 247, and 2047 are common tool limits).